File: N2.5/N2.5.cpp
#include <iostream>
#include <string>
#include <map>
#include <algorithm>
#include <vector>
#include <set>

using namespace std;

vector<string> WordConverter(const string& fourthblank) //func for Convert word from input string
{
    vector<string> ConvertedWord;
    string BlankWord = fourthblank;
    sort(BlankWord.begin(), BlankWord.end());
    ConvertedWord.push_back(BlankWord);
    while (next_permutation(BlankWord.begin(), BlankWord.end()))
    {
        ConvertedWord.push_back(BlankWord);
    }
    return ConvertedWord;
}

vector<string> StringSplitter(const string& EnterString) // func for split String
{
    vector<string> SplittedString;
    string MiniWord;
    for (int i = 0; i < EnterString.length(); i++)
    {
        if (EnterString[i] == ' ') 
        {
            SplittedString.push_back(MiniWord); 
            MiniWord = "";
            
        } else if (i == EnterString.length() - 1) 
        {
            MiniWord += EnterString[i];
            SplittedString.push_back(MiniWord); 
        }
        else 
        {
            MiniWord += EnterString[i];
        }
    }
    return SplittedString;
}

string Translate(string WordToTranslate, const map<string, string>& Dictionary) // func for Translate words to Morze
{
  string  TranslatedWord = "";
 for (int j = 0; j < WordToTranslate.length(); j++)
 {
    string fifthblank = WordToTranslate.substr(j, 1);
    TranslatedWord += Dictionary.at(fifthblank);
 }
 return TranslatedWord;
}

string UnicChecker(const vector<string>& TranslatedAllCombinations, string RNComb) // func for chech combination`s unic
{
    int SOS = 0; //ScoreOfSames
    for (string word : TranslatedAllCombinations)
    {
        if (RNComb == word) 
        {
            SOS += 1;
        }
    }
    if (SOS > 1)    
        {
            return ""; 
        }
    else
    {
        return RNComb;
    }
}

int main()
{   
    int checkofgo = 1;
    while (checkofgo)
    {
        map <string, string> Dictionary = {
            {"a", ".-"}, {"b", "-..."}, {"w", ".--"}, {"g", "--."}, {"d", "-.."}, 
            {"e", "."}, {"v", "...-"}, {"z", "--.."}, {"i", ".."}, {"j", ".---"}, 
            {"k", "-.-"}, {"l", ".-.."}, {"m", "--"}, {"n", "-."}, {"o", "---"}, 
            {"p", ".--."}, {"r", ".-."}, {"s", "..."}, {"t", "-"}, {"u", "..-"}, 
            {"f", "..-."}, {"h", "...."}, {"c", "-.-."}, {"q", "--.-"}, {"y", "-.--"}, {"x", "-..-"}
        };
        
        cout << "Enter String - ";
        string EnterString;
        getline(cin, EnterString);
        
        vector <string> SplittedString;
        for (string secblank : StringSplitter(EnterString))
        {
            SplittedString.push_back(secblank);
        }
        
        int BoolCheckOfWordLength = 1; 
        for (string thirdblank : SplittedString)
        {
            if (thirdblank.length() > 9) {
                BoolCheckOfWordLength = 0;
                break; 
            }
        }
        
        int BoolCheckOfStringLength = 1; 
        if (EnterString.length() > 100 || EnterString.length() < 1)
        {
            BoolCheckOfStringLength = 0;
        }
        
        if (!BoolCheckOfStringLength || !BoolCheckOfWordLength)
        {
            cout << "Please, enter correct string" << endl;
        }
        else
        {   
            vector<string> AllCombinations;
            for (string fourthblank : SplittedString)
            {
                for (string Combination : WordConverter(fourthblank))
                {
                    AllCombinations.push_back(Combination);
                }
            }
             vector<string> TranslatedAllCombinations;
            for (string WordToTranslate : AllCombinations)
            {
                TranslatedAllCombinations.push_back(Translate(WordToTranslate, Dictionary));
            }
            set<string> UnicCombinations;
            for (string RNComb : TranslatedAllCombinations)
            {
                string sixblank = UnicChecker(TranslatedAllCombinations, RNComb);
                if (sixblank == "")
                {
                    continue;
                }
                else
                {
                    UnicCombinations.insert(sixblank);
                }
            }
            cout << endl;
            cout << "All combinations and their Morze translation: " << endl;
            for (int outputcount = 0; outputcount < AllCombinations.size(); outputcount++)
            {
                cout << AllCombinations[outputcount] << " = " << "(" << TranslatedAllCombinations[outputcount] << ")" << endl;
            }
            cout << "Unic words: " << endl;
            for (string Output : UnicCombinations)
            {
                cout << "(" << Output << ")" << " ";
            }
            cout << endl << "____________________________________________________" << endl << "               " << "Value of unic words - " << UnicCombinations.size() << endl;
    }
    checkofgo = 0;
}
}

File: N2.5/N2.5.rs
use std::collections::{HashMap, HashSet};

fn word_converter(fourth_blank: &str) -> Vec<String> {
    let mut converted_word = Vec::new();
    let mut blank_word = fourth_blank.chars().collect::<Vec<_>>();
    blank_word.sort();
    converted_word.push(blank_word.iter().collect());
    while next_permutation(&mut blank_word) {
        converted_word.push(blank_word.iter().collect());
    }
    converted_word
}

fn next_permutation(arr: &mut [char]) -> bool {
    let mut i = arr.len() - 1;
    while i > 0 && arr[i - 1] >= arr[i] {
        i -= 1;
    }
    if i <= 0 {
        return false;
    }
    let mut j = arr.len() - 1;
    while arr[j] <= arr[i - 1] {
        j -= 1;
    }
    arr.swap(i - 1, j);
    arr[i..].reverse();
    true
}

fn string_splitter(enter_string: &str) -> Vec<String> {
    enter_string.split_whitespace().map(|s| s.to_string()).collect()
}

fn translate(word_to_translate: &str, dictionary: &HashMap<&str, &str>) -> String {
    let mut translated_word = String::new();
    for c in word_to_translate.chars() {
        if let Some(&morse) = dictionary.get(&c.to_string().as_str()) {
            translated_word.push_str(morse);
        }
    }
    translated_word
}

fn unic_checker(translated_all_combinations: &[String], rn_comb: &str) -> Option<String> {
    let mut sos = 0; // ScoreOfSames
    for word in translated_all_combinations {
        if rn_comb == word {
            sos += 1;
        }
    }
    if sos > 1 {
        None
    } else {
        Some(rn_comb.to_string())
    }
}

fn main() {
    let dictionary: HashMap<&str, &str> = [
        ("a", ".-"), ("b", "-..."), ("w", ".--"), ("g", "--."), ("d", "-.."),
        ("e", "."), ("v", "...-"), ("z", "--.."), ("i", ".."), ("j", ".---"),
        ("k", "-.-"), ("l", ".-.."), ("m", "--"), ("n", "-."), ("o", "---"),
        ("p", ".--."), ("r", ".-."), ("s", "..."), ("t", "-"), ("u", "..-"),
        ("f", "..-."), ("h", "...."), ("c", "-.-."), ("q", "--.-"), ("y", "-.--"), ("x", "-..-")
    ].iter().cloned().collect();
    
    let enter_string = "ign igg ezn smg"; // Введите вашу строку здесь
    
    let splitted_string = string_splitter(enter_string);
    
    let bool_check_of_word_length = splitted_string.iter().all(|word| word.len() <= 9);
    let bool_check_of_string_length = enter_string.len() <= 100 && enter_string.len() >= 1;
    
    if !bool_check_of_string_length || !bool_check_of_word_length {
        println!("Please, enter correct string");
    } else {
        let mut all_combinations = Vec::new();
        for fourth_blank in &splitted_string {
            for combination in word_converter(fourth_blank) {
                all_combinations.push(combination);
            }
        }
        let translated_all_combinations: Vec<String> = all_combinations
            .iter()
            .map(|word| translate(word, &dictionary))
            .collect();
        
        let mut unic_combinations = HashSet::new();
        for rn_comb in &translated_all_combinations {
            if let Some(six_blank) = unic_checker(&translated_all_combinations, rn_comb) {
                unic_combinations.insert(six_blank);
            }
        }
        
        println!();
        println!("All combinations and their Morze translation: ");
        for (output_count, combination) in all_combinations.iter().enumerate() { // В этой строке мы изменили образец шаблона для ссылки на элемент
            println!("{} = ({})", combination, translated_all_combinations[output_count]);
        }
        
        println!("Unic words: ");
        for output in &unic_combinations {
            println!("({})", output);
        }
        
        println!("____________________________________________________");
        println!("               Value of unic words - {}", unic_combinations.len());
    }
}
